### PR TITLE
OS-8071 pool vdevs should not have a probe failure SERD

### PR DESCRIPTION
**Pull request by KodyKantor**

### CODE REVIEW

File: usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c [hunk 1]
```diff
@@ -796,6 +827,39 @@ zfs_fm_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl, const char *class)
...
+	cb.cb_guid = pool_guid;
+	cb.cb_zhp = NULL;
+	if (zhdl != NULL && zpool_iter(zhdl, find_pool, &cb) != 0)
+		return;
+
+	zhp = cb.cb_zhp; /* NULL if pool was not found. */
+
+	/*
+	 * The libzfs API takes a string representation of a base-10 guid here
+	 * instead of a number, likely because the primary libzfs consumer are
```
Comment (rejohnst) [inline usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c]: s/consumer/consumers/

File: usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c [hunk 2]
```diff
@@ -796,6 +827,39 @@ zfs_fm_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl, const char *class)
...
+	 * the CLI tools.
+	 */
+	vdg[vdg_size - 1] = '\0';
+	vp = ulltostr(vdev_guid, &vdg[vdg_size - 1]);
```
Comment (rejohnst) [inline usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c]: I would just use snprintf here - it's also more portable anyways (ulltostr is a SVR4-ism)

File: usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c [hunk 3]
```diff
@@ -796,6 +827,39 @@ zfs_fm_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl, const char *class)
...
 
+	zhdl = fmd_hdl_getspecific(hdl);
+
+	/*
+	 * Find the corresponding pool.
+	 */
+	cb.cb_guid = pool_guid;
+	cb.cb_zhp = NULL;
+	if (zhdl != NULL && zpool_iter(zhdl, find_pool, &cb) != 0)
+		return;
```
Comment (rejohnst) [inline usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c]: Should we bump a stat or log a debug message for this error case?

Comment (KodyKantor) [reply usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c]: I added a stat for this. Does that look good? I think we should only hit this case when the pool iteration itself fails, which should be quite rare.

**Comment by KodyKantor** [comment]
@jjelinek, thanks for taking a look.
> 980 This conditional seems wrong. If isspare is false, the first test will evaluate to true, you will thus short-circuit the call to fmd_serd_record, but still set checkremove to true. Is this the intended behavior? Would using something like (isspare == 1 && fmd_serd_record...) be more correct?

I wrote out the three scenarios that we have to account for here and I think the conditional works correctly, but it is confusing. We do want to immediately set checkremove to true when isspare is false. I renamed `isspare` to `is_inactive_spare`, which might help make this more clear. `is_inactive_spare` is set to 0 when the spare is in use in the pool, or when the vdev is a 'normal' pool vdev. Both of these cases will short circuit the conditional and set checkremove without using SERD.

File: usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c (hunk 3 above)
Comment (rejohnst) [reply usr/src/cmd/fm/modules/common/zfs-diagnosis/zfs_de.c]: Yep, that looks good.   Thanks.

**Comment by KodyKantor** [comment]
We had some problems getting the bot to recognize the 'integration-approval' label so I used the 'force-integration-approval' label.

**Pull request merged**